File: src/test/java/cspfj/StatisticsManagerTest.java
package cspfj;

import static org.junit.Assert.assertEquals;

import org.junit.Before;
import org.junit.Test;

import cspfj.util.Statistic;

public final class StatisticsManagerTest {
    @SuppressWarnings("unused")
    private static class TestClass {
        @Statistic
        private static int testInt = 8;
        @Statistic
        private static long testLong = 8;
        @Statistic
        private static float testFloat = 9;
        @Statistic
        private static double testDouble = 9;

        @Statistic
        private int testDynInt;
        @Statistic
        private long testDynLong;
        @Statistic
        private float testDynFloat;
        @Statistic
        private double testDynDouble;

        public TestClass(int testDynInt, long testDynLong, float testDynFloat,
                double testDynDouble) {
            super();
            this.testDynInt = testDynInt;
            this.testDynLong = testDynLong;
            this.testDynFloat = testDynFloat;
            this.testDynDouble = testDynDouble;
        }
    }

    @Before
    public void setUp() throws Exception {
    }

    @Test
    public void testRegisterClassOfQ() {
        StatisticsManager.register(TestClass.class);

        assertEquals(8,
                StatisticsManager.get(TestClass.class.getName() + ".testInt"));
        assertEquals(9f,
                StatisticsManager.get(TestClass.class.getName() + ".testFloat"));
        assertEquals(8l,
                StatisticsManager.get(TestClass.class.getName() + ".testLong"));
        assertEquals(9d, StatisticsManager.get(TestClass.class.getName()
                + ".testDouble"));
    }

    @Test
    public void testRegisterStringObject() {
        StatisticsManager.register("test", new TestClass(8, 9, 10, 11));
        assertEquals(8, StatisticsManager.get("test.testDynInt"));
        assertEquals(9l, StatisticsManager.get("test.testDynLong"));
        assertEquals(10f, StatisticsManager.get("test.testDynFloat"));
        assertEquals(11d, StatisticsManager.get("test.testDynDouble"));
    }

    @Test
    public void testReset() {
        StatisticsManager.register(TestClass.class);
        StatisticsManager.register("test2", new TestClass(8, 9, 10, 11));
        StatisticsManager.reset();
        assertEquals(0,
                StatisticsManager.get(TestClass.class.getName() + ".testInt"));
        assertEquals(0l,
                StatisticsManager.get(TestClass.class.getName() + ".testLong"));
        assertEquals(0f,
                StatisticsManager.get(TestClass.class.getName() + ".testFloat"));
        assertEquals(0d, StatisticsManager.get(TestClass.class.getName()
                + ".testDouble"));
        assertEquals(4, StatisticsManager.digest().size());
    }

}

File: src/test/scala/cspfj/StatisticsManagerTest.scala
package cspfj;

import org.junit.Assert.assertEquals
import org.junit.Test

object StatisticsManagerTest {
  @Statistic
  val testInt = 8;
  @Statistic
  val testLong = 9l;
  @Statistic
  val testFloat = 10.0f;
  @Statistic
  val testDouble = 11.0;
}

final class StatisticsManagerTest {

  @Test
  def testRegisterStringObject() {
    val stats = new StatisticsManager
    stats.register("test", StatisticsManagerTest);
    assertEquals(8, stats("test.testInt"));
    assertEquals(9l, stats("test.testLong"));
    assertEquals(10f, stats("test.testFloat"));
    assertEquals(11d, stats("test.testDouble"));
  }

  @Test
  def testOps() {
    val seq = Seq(30, 53, 76, 77, 25, 66, 78, 50,
      70, 42, 90, 31, 79, 7, 39, 49, 15, 82, 35, 27)

    assertEquals(51.05, StatisticsManager.average(seq), 0)
    assertEquals(50, StatisticsManager.median(seq), 0)
    assertEquals(24.81187, StatisticsManager.stDev(seq), 0.00001)
  }

}
